File: src/main/java/org/kite9/diagram/visualization/compaction/PluggableCompactor.java
package org.kite9.diagram.visualization.compaction;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.kite9.diagram.common.elements.PositionAction;
import org.kite9.diagram.common.elements.vertex.Vertex;
import org.kite9.diagram.model.Rectangular;
import org.kite9.diagram.model.position.Direction;
import org.kite9.diagram.visualization.compaction.segment.Segment;
import org.kite9.diagram.visualization.compaction.segment.SegmentBuilder;
import org.kite9.diagram.visualization.orthogonalization.Dart;
import org.kite9.diagram.visualization.orthogonalization.DartFace;
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization;
import org.kite9.framework.common.HelpMethods;


/**
 * Superclass for all segment-based compaction methods.   The process works by converting each face into
 * a set of segments, where segments represent horizontal or vertical positions.  
 *  
 * 
 * @author robmoffat
 *
 */
public class PluggableCompactor implements Compactor {

	public PluggableCompactor(CompactionStep[] steps) {
		this.steps = steps;
	}
	
	protected SegmentBuilder sb = new SegmentBuilder();
	protected CompactionStep[] steps;

	public static final Set<Direction> VERTICAL = HelpMethods.createSet(Direction.UP, Direction.DOWN);
	public static final Set<Direction> HORIZONTAL = HelpMethods.createSet(Direction.LEFT, Direction.RIGHT);
	
	public Compaction compactDiagram(Orthogonalization o) {
		List<Segment> horizontal = buildSegmentList(o, HORIZONTAL);
		List<Segment> vertical = buildSegmentList(o, VERTICAL);
		Map<Dart, Segment> dartToSegmentMap = calculateDartToSegmentMap(horizontal, vertical);
		Map<Vertex, Segment> horizontalSegmentMap = createVertexSegmentMap(horizontal);
		Map<Vertex, Segment> verticalSegmentMap = createVertexSegmentMap(vertical);
		Compaction compaction = instantiateCompaction(o, horizontal, vertical, dartToSegmentMap, horizontalSegmentMap, verticalSegmentMap);
		compact(o.getPlanarization().getDiagram(), compaction);
		return compaction;
	}

	protected Compaction instantiateCompaction(Orthogonalization o, List<Segment> horizontal, List<Segment> vertical, Map<Dart, Segment> dartToSegmentMap,
			Map<Vertex, Segment> horizontalSegmentMap, Map<Vertex, Segment> verticalSegmentMap) {
		return new CompactionImpl(o, horizontal, vertical, horizontalSegmentMap, verticalSegmentMap, dartToSegmentMap);
	}
	
	private Map<Dart, Segment> calculateDartToSegmentMap(List<Segment> h1, List<Segment> v1) {
		Map<Dart, Segment> out = new HashMap<>();
		h1.stream().forEach(s -> addSegmentsToMap(out, s));
		v1.stream().forEach(s -> addSegmentsToMap(out, s));
		return out;
	}
	
	private void addSegmentsToMap(Map<Dart, Segment> dartSegmentMap, Segment segment) {
		for (Dart d : segment.getDartsInSegment()) {
			dartSegmentMap.put(d, segment);
		}
	}

	@Override
	public void compact(Rectangular r, Compaction c) {
		for (CompactionStep step : steps) {
			step.compact(c, r, this);
		}
	}
	
	protected Map<Vertex, Segment> createVertexSegmentMap(List<Segment> segs) {
		Map<Vertex, Segment> vertexToSegmentMap = new HashMap<Vertex, Segment>();
		for (Segment segment : segs) {
			for (Vertex v : segment.getVerticesInSegment()) {
				vertexToSegmentMap.put(v, segment);
			}
		}
		return vertexToSegmentMap;
	}

	public List<Segment> buildSegmentList(Orthogonalization o, Set<Direction> direction) {
		List<Segment> segments = sb.buildSegmentList(o, direction, direction==HORIZONTAL ? PositionAction.YAction : PositionAction.XAction);
		return segments;
	}
	
}

File: src/main/kotlin/org/kite9/diagram/visualization/compaction/PluggableCompactor.kt
package org.kite9.diagram.visualization.compaction

import org.kite9.diagram.common.elements.Dimension
import org.kite9.diagram.common.elements.vertex.Vertex
import org.kite9.diagram.logging.LogicException
import org.kite9.diagram.model.position.Direction
import org.kite9.diagram.visualization.compaction.segment.Segment
import org.kite9.diagram.visualization.compaction.segment.SegmentBuilder
import org.kite9.diagram.visualization.orthogonalization.Dart
import org.kite9.diagram.visualization.orthogonalization.DartFace
import org.kite9.diagram.visualization.orthogonalization.Orthogonalization

/**
 * Superclass for all segment-based compaction methods.   The process works by converting each face into
 * a set of segments, where segments represent horizontal or vertical positions.
 *
 *
 * @author robmoffat
 */
class PluggableCompactor(protected var steps: Array<CompactionStep>) : Compactor {

    protected var sb = SegmentBuilder()

    override fun compactDiagram(o: Orthogonalization): Compaction {
        val horizontal = buildSegmentList(o, HORIZONTAL)
        val vertical = buildSegmentList(o, VERTICAL)
        val dartToSegmentMap = calculateDartToSegmentMap(horizontal, vertical)
        val horizontalSegmentMap = createVertexSegmentMap(horizontal)
        val verticalSegmentMap = createVertexSegmentMap(vertical)
        val topEmbedding = generateEmbeddings(o)
        val compaction = instantiateCompaction(
            o,
            horizontal,
            vertical,
            dartToSegmentMap,
            horizontalSegmentMap,
            verticalSegmentMap,
            topEmbedding
        )
        compact(compaction.getTopEmbedding(), compaction)
        return compaction
    }

    private fun generateEmbeddings(o: Orthogonalization): Embedding {
        val done: MutableMap<DartFace, EmbeddingImpl> = LinkedHashMap()
        var embeddingNumber = 0
        for (dartFace in o.getFaces()) {
            if (!done.containsKey(dartFace)) {
                var touching: MutableSet<DartFace> = HashSet()
                touching = getTouchingFaces(touching, dartFace, o)
                val ei = EmbeddingImpl(embeddingNumber++, ArrayList(touching))
                touching.forEach { df: DartFace -> done[df] = ei }
            }
        }
        for (e in LinkedHashSet(done.values)) {
            val contained =
                e.dartFaces
                    .flatMap { it.containedFaces }
            val inside: List<Embedding> = contained
                .map { df: DartFace -> done[df]!! }
            e.innerEmbeddings = inside
        }

        // return just the top one
        val topEmbedding = done.values.stream().filter { e: EmbeddingImpl -> isTopEmbedding(e) }
            .findFirst().orElseThrow { LogicException("No top embedding") }
        topEmbedding.isTopEmbedding = true
        return topEmbedding
    }

    private fun isTopEmbedding(e: Embedding): Boolean {
        return e.dartFaces.stream().anyMatch { df: DartFace -> df.outerFace && df.getContainedBy() == null }
    }

    private fun getTouchingFaces(
        touching: MutableSet<DartFace>,
        dartFace: DartFace,
        o: Orthogonalization
    ): MutableSet<DartFace> {
        var touching = touching
        val todo = touching.add(dartFace)
        if (todo) {
            for ((dart) in dartFace.dartsInFace) {
                val faces = o.getDartFacesForDart(dart)
                for (f2 in faces!!) {
                    if (f2 !== dartFace) {
                        touching = getTouchingFaces(touching, f2, o)
                    }
                }
            }
        }
        return touching
    }

    protected fun instantiateCompaction(
        o: Orthogonalization,
        horizontal: List<Segment>,
        vertical: List<Segment>,
        dartToSegmentMap: Map<Dart, Segment>,
        horizontalSegmentMap: Map<Vertex, Segment>,
        verticalSegmentMap: Map<Vertex, Segment>,
        topEmbedding: Embedding
    ): Compaction {
        return CompactionImpl(
            o,
            horizontal,
            vertical,
            horizontalSegmentMap,
            verticalSegmentMap,
            dartToSegmentMap,
            topEmbedding
        )
    }

    private fun calculateDartToSegmentMap(h1: List<Segment>, v1: List<Segment>): Map<Dart, Segment> {
        val out: MutableMap<Dart, Segment> = HashMap()
        h1.stream().forEach { s: Segment -> addSegmentsToMap(out, s) }
        v1.stream().forEach { s: Segment -> addSegmentsToMap(out, s) }
        return out
    }

    private fun addSegmentsToMap(dartSegmentMap: MutableMap<Dart, Segment>, segment: Segment) {
        for (d in segment.dartsInSegment) {
            dartSegmentMap[d] = segment
        }
    }

    override fun compact(e: Embedding, c: Compaction) {
        for (step in steps) {
            step.compact(c, e, this)
        }
    }

    protected fun createVertexSegmentMap(segs: List<Segment>): Map<Vertex, Segment> {
        val vertexToSegmentMap: MutableMap<Vertex, Segment> = HashMap()
        for (segment in segs) {
            for (v in segment.getVerticesInSegment()) {
                vertexToSegmentMap[v] = segment
            }
        }
        return vertexToSegmentMap
    }

    fun buildSegmentList(
        o: Orthogonalization,
        direction: Set<Direction>
    ): List<Segment> {
        return sb.buildSegmentList(o, direction, if (direction === HORIZONTAL) Dimension.H else Dimension.V)
    }

    companion object {
        val VERTICAL = setOf<Direction>(Direction.UP, Direction.DOWN)
        val HORIZONTAL = setOf<Direction>(Direction.LEFT, Direction.RIGHT)
    }
}